activerecord/lib/active_record/inheritance.rb

```diff
@@ -135,19 +135,28 @@ def sti_name
         # Returns the class type of the record using the current module as a prefix. So descendants of
         # MyApp::Business::Account would appear as MyApp::Business::AccountSubclass.
         def compute_type(type_name)
-          if type_name.match(/^::/)
+          if type_name.start_with?("::".freeze)
             # If the type is prefixed with a scope operator then we assume that
             # the type_name is an absolute reference.
             ActiveSupport::Dependencies.constantize(type_name)
           else
+            @_type_candidates_cache ||= Concurrent::Map.new
```

Comment:
I think this `||=` is technically not threadsafe - is that right @matthewd? Technically in that it's fine on MRI but not jRuby - use an `inherited` hook to initialize it instead.